kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ConstantSaslIT.java

```diff
@@ -0,0 +1,119 @@
+                ConstantSasl.class.getName(),
+                ConstantSasl.class.getName())
+                .withConfig(filterConfig)
+                .build();
+        NamedFilterDefinition lawyer = new NamedFilterDefinitionBuilder(
+                ClientTlsAwareLawyer.class.getName(),
+                ClientTlsAwareLawyer.class.getName())
+                .build();
+        var config = proxy(cluster)
+                .addToFilterDefinitions(saslInspection, lawyer)
+                .addToDefaultFilters(saslInspection.name(), lawyer.name());
+
+        try (var tester = kroxyliciousTester(config);
+                var producer = tester.producer();
+                var consumer = tester
+                        .consumer(Serdes.String(), Serdes.ByteArray(), Map.of(
+                                ConsumerConfig.GROUP_ID_CONFIG, "my-group-id",
+                                ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest",
+                                ConsumerConfig.FETCH_MAX_BYTES_CONFIG, "1"))) {
+
+            assertThat(producer.send(new ProducerRecord<>(topic.name(), "my-key", "my-value")))
+                    .succeedsWithin(Duration.ofSeconds(5));
+
+            producer.flush();
+
+            consumer.subscribe(Set.of(topic.name()));
+            var records = consumer.poll(Duration.ofSeconds(10));
+
+            assertThat(records).hasSize(1);
+            var recordHeaders = assertThat(records.records(topic.name()))
+                    .as("topic %s records", topic.name())
+                    .singleElement()
+                    .asInstanceOf(new InstanceOfAssertFactory<>(ConsumerRecord.class, KafkaAssertions::assertThat))
+                    .headers();
+            recordHeaders.singleHeaderWithKey(ClientAuthAwareLawyerFilter.HEADER_KEY_CLIENT_SASL_CONTEXT_PRESENT)
+                    .value().containsExactly(expectedClientSaslPresent ? (byte) 1 : (byte) 0);
+            recordHeaders.singleHeaderWithKey(ClientAuthAwareLawyerFilter.HEADER_KEY_CLIENT_SASL_AUTHORIZATION_ID)
+                    .hasValueEqualTo(expectedAuthorizedId);
+            recordHeaders.singleHeaderWithKey(ClientAuthAwareLawyerFilter.HEADER_KEY_CLIENT_SASL_MECH_NAME)
+                    .hasValueEqualTo(expectedMechanism);
+
+        }
+    }
+
+    @Test
+    void shouldHavePresentClientSaslContextWhenAuthSuccessful(
+                                                              KafkaCluster cluster,
+                                                              Topic topic) {
+        String mechanism = "FOO";
+        String authorizedId = "bob";
+        Map<String, Object> filterConfig = Map.of("api", ApiKeys.PRODUCE,
+                "mechanism", mechanism,
+                "authorizedId", authorizedId);
+        doAThing(cluster, topic, filterConfig, true, authorizedId, mechanism);
+    }
+
+    @Test
+    void shouldHaveEmptyClientSaslContextWhenAuthFailed(
```

Comment:
could also potentially add a check that:

1. request 1 -> success
2. request 2 -> fail

results in an empty context

Reply:
edit: and check the inverse, to exercise the interleaving behaviours

Reply:
I don't think this is possible right now, because the `ConstantSaslFilter` only calls `context.clientSaslAuthentication(Success|Failure)` once (hence the constant part of the name). We could change that by allowing the configuration the ability to specify multiple different calls to make, but I think we have unit tests for the interleaving, so I'm inclined to leave it for now.